File: api-go/internal/db/neo4j.go
package db

import (
	"context"
	"log"

	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

var Neo4jDriver neo4j.DriverWithContext

func InitNeo4j(uri, user, password string) error {
	var err error
	Neo4jDriver, err = neo4j.NewDriverWithContext(uri, neo4j.BasicAuth(user, password, ""))
	if err != nil {
		return err
	}

	if err := Neo4jDriver.VerifyConnectivity(context.Background()); err != nil {
		log.Printf("[neo4j] Warning: connectivity check failed: %v", err)
		return nil
	}

	log.Println("[neo4j] Connected")
	return nil
}

func InitNeo4jSchema(ctx context.Context) {
	session := Neo4jDriver.NewSession(ctx, neo4j.SessionConfig{})
	defer session.Close(ctx)

	queries := []string{
		"CREATE CONSTRAINT user_id IF NOT EXISTS FOR (u:User) REQUIRE u.id IS UNIQUE",
		"CREATE CONSTRAINT post_id IF NOT EXISTS FOR (p:Post) REQUIRE p.id IS UNIQUE",
		"CREATE INDEX user_username IF NOT EXISTS FOR (u:User) ON (u.username)",
	}

	for _, q := range queries {
		if _, err := session.Run(ctx, q, nil); err != nil {
			log.Printf("[neo4j] Schema warning: %v", err)
		}
	}
	log.Println("[neo4j] Schema initialized")
}

func CloseNeo4j(ctx context.Context) {
	if Neo4jDriver != nil {
		Neo4jDriver.Close(ctx)
	}
}

File: twitclone-api/src/utils/neo4j.ts
import neo4j, { type Driver, type Session } from "neo4j-driver";

let driver: Driver | null = null;

export function getNeo4jDriver(): Driver {
    if (!driver) {
        const uri = process.env.NEO4J_URI || "bolt://localhost:7687";
        const user = process.env.NEO4J_USER || "neo4j";
        const password = process.env.NEO4J_PASSWORD || "neo4jdev";
        driver = neo4j.driver(uri, neo4j.auth.basic(user, password));
    }
    return driver;
}

export function getSession(): Session {
    return getNeo4jDriver().session();
}

export async function initNeo4jSchema(): Promise<void> {
    const session = getSession();
    try {
        await session.run(
            "CREATE CONSTRAINT user_id IF NOT EXISTS FOR (u:User) REQUIRE u.id IS UNIQUE"
        );
        await session.run(
            "CREATE CONSTRAINT post_id IF NOT EXISTS FOR (p:Post) REQUIRE p.id IS UNIQUE"
        );
        await session.run(
            "CREATE INDEX user_username IF NOT EXISTS FOR (u:User) ON (u.username)"
        );
        console.log("[neo4j] Schema constraints and indexes initialized");
    } catch (err) {
        console.error("[neo4j] Failed to initialize schema:", err);
    } finally {
        await session.close();
    }
}

export async function closeNeo4j(): Promise<void> {
    if (driver) {
        await driver.close();
        driver = null;
    }
}
